lib/Common/Memory/SmallNormalHeapBucket.cpp

```diff
@@ -401,17 +448,59 @@ SmallNormalHeapBucketBase<TBlockType>::SweepPartialReusePages(RecyclerSweep& rec
 #if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP && SUPPORT_WIN32_SLIST && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP_USE_SLIST
         if (isReused)
         {
-            if (this->AllowAllocationsDuringConcurrentSweep())
-            {
-                bool blockAddedToSList = HeapBucketT<TBlockType>::PushHeapBlockToSList(this->allocableHeapBlockListHead, heapBlock);
+            DebugOnly(heapBlock->blockNotReusedInPartialHeapBlockList = false);
 
-                // If we encountered OOM while pushing the heapBlock to the SLIST we must add it to the heapBlockList so we don't lose track of it.
-                if (!blockAddedToSList)
+            DebugOnly(this->AssertCheckHeapBlockNotInAnyList(heapBlock));
+            if (heapBlock->HasFreeObject())
+            {
+                if (this->AllowAllocationsDuringConcurrentSweep())
                 {
-                    heapBlock->SetNextBlock(this->heapBlockList);
-                    this->heapBlockList = heapBlock;
+                    Assert(!this->IsAnyFinalizableBucket());
+                    bool blockAddedToSList = HeapBucketT<TBlockType>::PushHeapBlockToSList(this->allocableHeapBlockListHead, heapBlock);
+
+                    // If we encountered OOM while pushing the heapBlock to the SLIST we must add it to the heapBlockList so we don't lose track of it.
+                    if (!blockAddedToSList)
+                    {
+#ifdef RECYCLER_TRACE
+                        this->GetRecycler()->PrintBlockStatus(this, heapBlock, _u("[**10**] finished SweepPartialReusePages, heapblock REUSED but OOM while adding to the SLIST."));
+#endif
+                        this->GetRecycler()->OutOfMemory();
```

Comment:
You can't logically continue at this point.  OOM may not always FailFast. 

We should find a way to recover from this case just like before.

Reply:
Ok. I will change this to fail fast for now and will plan to work on the correct thing to do here later.

---
In reply to: [148566503](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148566503) [](ancestors = 148566503)